edx_solutions_api_integration/courses/views.py

```diff
@@ -2381,33 +2381,22 @@ def get(self, request, course_id):  # pylint: disable=W0613
         if not course_exists(course_id):
             raise Http404
 
-        response_data = []
+        user_id = self.request.query_params.get('user_id', None)
+        role = self.request.query_params.get('role', None)
         course_key = get_course_key(course_id)
         instructors = CourseInstructorRole(course_key).users_with_role()
-        for instructor in instructors:
-            response_data.append({'id': instructor.id, 'role': 'instructor'})
-
         staff = CourseStaffRole(course_key).users_with_role()
-        for admin in staff:
-            response_data.append({'id': admin.id, 'role': 'staff'})
-
         observers = CourseObserverRole(course_key).users_with_role()
-        for observer in observers:
-            response_data.append({'id': observer.id, 'role': 'observer'})
-
         assistants = CourseAssistantRole(course_key).users_with_role()
-        for assistant in assistants:
-            response_data.append({'id': assistant.id, 'role': 'assistant'})
-
-        user_id = self.request.query_params.get('user_id', None)
+        all_users = (instructors | staff | observers | assistants).annotate(role=F("courseaccessrole__role"))
         if user_id:
-            response_data = list([item for item in response_data if int(item['id']) == int(user_id)])
-
-        role = self.request.query_params.get('role', None)
+            all_users &= User.objects.filter(id=int(user_id))
         if role:
-            response_data = list([item for item in response_data if item['role'] == role])
-
-        return Response(response_data, status=status.HTTP_200_OK)
+            all_users &= User.objects.filter(courseaccessrole__role=role)
```

Comment:
```suggestion
            all_users &= User.objects.filter(role=role)
```
You should filter by the annotation here otherwise you'll end up with the situation described below.

Reply:
I had tried this before and it didn't work, but that may have been when I was annotating later in the chain. I'll try again now that I annotate earlier.

Reply:
I tested it to be sure since I was not entirely sure as well, but it does work if you annotate first and then filter.

Reply:
So I realized you meant annotate on the right-hand QuerySet, because it still didn't work for me if I relied on the annotation done earlier on the left-hand QuerySet (all_users). I'm not sure if that's worth it though, because it seems to have not changed the raw SQL at all, literally. This was the SQL before and after:

```sql
SELECT DISTINCT `auth_user`.`id`,
                `student_courseaccessrole`.`role` AS `role`
FROM `auth_user`
INNER JOIN `student_courseaccessrole` ON (`auth_user`.`id` = `student_courseaccessrole`.`user_id`)
LEFT OUTER JOIN `student_courseaccessrole` T3 ON (`auth_user`.`id` = T3.`user_id`)
WHERE (((`student_courseaccessrole`.`course_id` = course-v1:edX+DemoX+Demo_Course
         AND `student_courseaccessrole`.`role` = instructor
         AND `student_courseaccessrole`.`org` = edX)
        OR (`student_courseaccessrole`.`course_id` = course-v1:edX+DemoX+Demo_Course
            AND `student_courseaccessrole`.`role` = staff
            AND `student_courseaccessrole`.`org` = edX)
        OR (`student_courseaccessrole`.`course_id` = course-v1:edX+DemoX+Demo_Course
            AND `student_courseaccessrole`.`role` = observer
            AND `student_courseaccessrole`.`org` = edX)
        OR (`student_courseaccessrole`.`course_id` = course-v1:edX+DemoX+Demo_Course
            AND `student_courseaccessrole`.`role` = assistant
            AND `student_courseaccessrole`.`org` = edX))
       AND T3.`role` = assistant)
```

Reply:
In any case that extra join on `T3` is definitely incorrect... I wouldn't think the ORM was going to do it that way, and that's what's causing at least a part of the duplication. It needs to be checking `student_courseaccessrole.role = assistant`, not `T3.role = assistant`.